### DISCUSSION TITLE
calculating transmission latency

### DISCUSSION THREAD

**Discussion by bazookam7** [google/oboe#1130] [Q&A] [answered]
The function getCurrentOutputLatencyMillis shows the average latency of the audio stream between data entering the stream and it being presented to the audio device so basically it is buffering latency + transmission latency right?
What i mean by transmission latency is the time a particular frame takes after coming out of aaudio stream buffer to reach the presentation layer and
what i mean by buffering latency is the time a particular frame has to stay inside the buffer before leaving.
Suppose buffer size = 2*burst size , how to measure transmission latency in this case using aaudio/obeo library or can we get the point where a chunk of data is just going out of buffer ?
or how to measure buffering latency?

**Comment by philburk** [comment, marked as answer]
Your description of the latency is correct.

> Suppose buffer size = 2*burst size , how to measure transmission latency in this case

You can measure the amount of data currently in the buffer by subtracting:

        framesInBuffer = stream->getFramesWritten() - stream->getFramesRead();
        bufferLatencyMillis = 1000 * framesInBuffer / sampleRate;
        transmissionLatencyMillis = measuredLatencyMillis - bufferLatencyMillis;

But please do not call getFramesRead() from a callback. On older systems, that can cause a race condition.
https://github.com/google/oboe/blob/master/docs/notes/rlsbuffer.md

It is OK to call it from another thread outside the callback.

Also note that the latency will jump up every time you write into the buffer and then slowly drop as the audio is played out the speaker. So it looks like a sawtooth:  |\|\|\

**Comment by philburk** [comment]
(Since people have more questions, I am converting this to a discussion.)

> So framesInBuffer above should be equal to AAudioStream_getBufferSizeInFrames(stream) every time ?

No. AAudioStream_getBufferSizeInFrames() tells you the maximum amount of data that can be in the buffer.
The actual amount will be between 0 and size. The BufferSize determines the maximum latency for that buffer.
If the BufferSize is too low then it is more likely to underflow and cause glitches.
You can adjust BufferSize to trade latency for glitches.
 
AAudioStream_getBufferCapacityInFrames() is the amount of memory allocated for the buffer.
That determines the upper limit for BufferSize.

**Comment by bazookam7** [comment]
can we write using aaudio and capture at lower layer via ALSA loopback?

**Comment by philburk** [comment]
No. ALSA loopback is not available on Android. There are privacy issues and the HAL hides ALSA.

**Comment by darth07** [comment]
Can we use these streams for streaming over rtp ? (after converting it to AudioStream which carrys audio payloads over RTP because mediarecorder use Audiostream(rtp stream) for streaming)
